### PR TITLE
feat: 新增 Cluster/Pod 级 DNS Server 故障注入场景

### PR DESCRIPTION
**Pull request by tdy218**
### Describe what this PR does / why we need it

在 chaosblade-operator 中新增两个 DNS Server 故障注入场景，分别覆盖 Cluster 与 Pod 两个维度，并在销毁实验时自动恢复原状：

1. **Cluster 级 DNS Server 故障**：根据 K8s 集群 DNS Service 地址（默认 `kube-dns` / `kube-system`，兼容 CoreDNS）反推出 DNS Server 工作负载，将其副本数缩容至 0，使整个集群的 DNS 解析完全不可用。
2. **Pod 级 DNS Server 故障**：根据指定的 Pod 名称（以及可选的 nameserver IP）查找该 Pod 的 DNS 配置，通过覆写其所属工作负载（Deployment/DaemonSet/StatefulSet）的 PodSpec，注入不可达的 nameserver `127.0.0.255`，使该 Pod 视角下的 DNS 解析完全不可用。

同步更新中英文 README 的 `Supported experiments` / `支持的场景` 章节，并提供示例 CR YAML。

### Does this pull request fix one issue?

NONE

#### 1. 新增 `cluster` Scope（`exec/cluster/`）

- `cluster.go`：注册 `cluster` Scope 下的 `dns` 目标。
- `controller.go`：实现 `ExpController`，与现有 `pod` / `node` / `container` / `service` 的风格一致。
- `dnsfailure.go`：实现 `dnsfailure` Action：
  - 通过 `--dns-service` / `--dns-service-namespace` 找到集群 DNS Service。
  - 读取 `Service.Spec.Selector`，与同命名空间下的 Deployment 的 PodTemplate Labels 进行匹配，**反推出 DNS Server 工作负载**。
  - 使用 `RetryOnConflict` 包裹更新；将原始 `replicas` 序列化后存入 annotation，并将 `replicas` 缩容为 0。
  - 销毁时按实验 ID 校验 annotation，恢复原始副本数并清理 annotation。
- 在 `exec/controller.go` 中通过 `cluster.NewExpController(client)` 注册到分发器。

#### 2. 新增 Pod 级 `dnsfailure` Action（`exec/pod/dnsfailureexp.go`）

- 复用 Pod scope 既有的 `--names` / `--labels` / `--namespace` 选择能力，并附加可选的 `--nameserver` IP 校验。
- 沿 `OwnerReferences` 向上解析（`ReplicaSet → Deployment` / `DaemonSet` / `StatefulSet`）找到顶层工作负载。
- 备份 `PodTemplate.Spec.DNSPolicy` 与 `DNSConfig`（JSON 编码）至工作负载 annotation；将 `DNSPolicy` 改为 `None`，`DNSConfig.Nameservers` 设为不可路由的回环 `127.0.0.255`。
- 删除 Pod 触发控制器重建；多 Pod 共享同一 workload 时通过 `processedWorkloads` 去重。
- 销毁时优先按 Pod 反查 owner 走精确恢复路径；当 Pod 已不存在时回退按 namespace + 实验 ID annotation 枚举三类工作负载（Deployment/DaemonSet/StatefulSet）全量恢复。
- 在 `exec/pod/pod.go` 的 `NewSelfExpModelCommandSpec` 中注册，并补充 LongDesc 与 Example。

#### 3. 文档与示例

- `README.md` / `README_CN.md` 在 `Supported experiments` 章节：
  - Pod 类追加 DNS 项；
  - 末尾新增 `Cluster` 大类与 `DNS` 项。
- 新增 `examples/cluster-dns-failure.yaml`、`examples/pod-dns-failure.yaml`，可直接 `kubectl apply` 验证。

#### 本地构建 / 检查

```bash
go build ./...
go vet ./...
go test ./exec/pod/... ./exec/cluster/...
```

#### 集群 E2E（建议在测试集群执行）

1. 部署本变更后的 chaosblade-operator。
2. **Cluster 级 DNS 故障**：

```bash
kubectl apply -f examples/cluster-dns-failure.yaml
# 预期：kube-system/coredns（或 kube-dns）Deployment 副本数被缩为 0，
# 在任意 Pod 中 nslookup kubernetes.default 应失败。
kubectl delete -f examples/cluster-dns-failure.yaml
# 预期：Deployment 副本数恢复，集群 DNS 解析恢复正常。
```

3. **Pod 级 DNS 故障**：

```bash
kubectl apply -f examples/pod-dns-failure.yaml
# 预期：目标 Pod 重建后，进入新 Pod 执行 nslookup 失败；
# 工作负载 PodSpec 上可见 chaosblade.io/* 相关 annotation。
kubectl delete -f examples/pod-dns-failure.yaml
# 预期：工作负载 DNSPolicy/DNSConfig 恢复，重建后的 Pod DNS 解析正常。
```

#### chaosblade CLI

```bash
# Cluster
blade create k8s cluster-dns dnsfailure --kubeconfig ~/.kube/config

# Pod
blade create k8s pod-pod dnsfailure --names nginx-app-xxx --namespace default --kubeconfig ~/.kube/config
```

### Special notes for reviews

- 故障的恢复语义遵循现有 ChaosBlade Operator 的 reconciliation 模型，通过 `kubectl delete blade <name>` 触发 destroy 实现恢复，与 `schedulingfailure`、`badresourcesize` 等 Action 保持一致。
- Cluster Scope 为新增的顶层 Scope（与 `pod` / `node` / `container` / `service` 并列），后续可承接更多集群级故障（如 etcd、kube-proxy 等）。
- 所有写操作使用 `RetryOnConflict` 包裹，并通过 `chaosblade.io/experiment` annotation 防止跨实验冲突；幂等：相同实验 ID 二次注入无副作用。
- RBAC：当前 `deploy/helm/chaosblade-operator/templates/rbac.yaml` 已对 `services`、`deployments`、`daemonsets`、`statefulsets`、`pods` 授予 `*` 权限，本次实现未引入新的 API 资源，无需调整 RBAC。
- 注入 Cluster 级 DNS 故障会影响整个集群的服务发现（包括 operator 自身可能依赖的 DNS 解析），请仅在测试环境使用。

Signed-off-by: tdy218 <tdy218@gmail.com>

### CODE REVIEW

File: exec/pod/dnsfailureexp.go [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+
+	annotations[ChaosBladePodDnsFailureAnnotation] = ChaosBladePodDnsFailureAction
+	annotations[ChaosBladeExperimentAnnotation] = experimentId
+
+	podSpec.DNSPolicy = v1.DNSNone
+	podSpec.DNSConfig = &v1.PodDNSConfig{
+		Nameservers: []string{UnreachableDnsNameserver},
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (tdy218) [reply exec/pod/dnsfailureexp.go:575]: 已经增加测试用例

**Review by oss-sentinel-ai — COMMENTED** [review]
## Code Review: Cluster/Pod DNS Server Fault Injection

This PR adds two new DNS fault injection scenarios at cluster and pod level. The implementation is well-structured with good test coverage (882 + 1446 lines of tests). Here are my findings:

### Strengths
- Comprehensive test coverage with unit tests for both cluster and pod DNS failure scenarios
- Good use of annotations to track experiment state and enable clean rollback
- Proper error handling with actionable messages (e.g., suggesting `--dns-deployment` flag when ambiguous)
- The `findDnsDeployment` function correctly handles the ambiguous case (multiple Deployments matching DNS Service selector)

### Issues Found

#### 🔴 Critical: Pod DNS failure modifies workload spec without owner reference check
In `exec/pod/dnsfailureexp.go`, the `executePodDnsFailure` function modifies the parent workload (Deployment/DaemonSet/StatefulSet) PodSpec directly. If the experiment controller crashes mid-experiment, the annotation `chaosblade.io/pod-dnsfailure` may exist on the workload but the ChaosBlade CR status won't reflect the experiment state, making cleanup ambiguous.

**Recommendation:** Consider adding a finalizer or using the existing ChaosBlade CR status as the source of truth for cleanup decisions, rather than relying solely on workload annotations.

#### 🟡 Medium: Hardcoded nameserver `127.0.0.255` should be configurable
In `exec/pod/dnsfailureexp.go`, the injected nameserver is hardcoded as `127.0.0.255`. While this is a reasonable default (loopback address that won't respond), some environments may need a different unreachable address.

**Recommendation:** Consider exposing this as an optional flag (e.g., `--dns-failure-nameserver`) with `127.0.0.255` as the default.

#### 🟡 Medium: `retry.RetryOnConflict` without backoff in cluster DNS
In `exec/cluster/dnsfailure.go`, the `retry.RetryOnConflict` calls use default retry parameters. For cluster-critical operations (scaling DNS to 0 replicas), a more conservative retry strategy with exponential backoff would be safer.

#### 🟢 Minor: Test file size
The test files are very large (882 + 1446 lines). Consider splitting into sub-tests by scenario (create/destroy/edge-cases) for better maintainability.

### Overall Assessment
The PR is well-designed with thorough testing. The critical issue should be addressed before merge to prevent orphaned workload modifications on controller crash. The other suggestions are improvements that can be addressed in follow-up PRs.

**Comment by xcaspar** [comment]
## 🧪 Automated Test Report — PR #303

**67 tests all passing** (37 PR-authored + 30 supplemental)

| Category | Tests | Pass | Fail | Fix Rounds |
|----------|-------|------|------|------------|
| PR tests (cluster) | 12 | 12 | 0 | — |
| PR tests (pod DNS) | 25 | 25 | 0 | — |
| Supplemental (cluster) | 14 | 14 | 0 | 1 |
| Supplemental (pod DNS) | 16 | 16 | 0 | 0 |
| **Total** | **67** | **67** | **0** | — |

### PR Test Quality Assessment

Excellent. The PR ships **2,328 lines of tests** covering all core business logic, boundary conditions, and E2E lifecycles. Supplemental tests focus on gaps: registration/wiring verification, additional edge cases, DaemonSet/StatefulSet round trips, and multi-pod deduplication.

### ⚠️ Finding: `readOriginalReplicas` int32 Overflow (Low Risk)

**Location**: `exec/cluster/dnsfailure.go:385-397`

`readOriginalReplicas("2147483648")` silently wraps to `-2147483648` instead of returning an error. Root cause: `strconv.Atoi` on 64-bit systems parses `2147483648` as a valid `int64`, then `int32(n)` truncates/wraps.

**Impact**: Low risk — K8s Deployment replicas won't approach int32 max in practice, and `json.Marshal(int32)` always produces in-range values. However, manually edited annotations with out-of-range values would corrupt the restore replica count.

**Suggestion**: Add a range check after `strconv.Atoi` succeeds:
```go
if n > math.MaxInt32 || n < math.MinInt32 {
    return 0, fmt.Errorf("replicas %d out of int32 range", n)
}
```

<details>
<summary><strong>Supplemental Cluster Tests (14 functions)</strong></summary>

**Registration/wiring (4):**
- `TestNewResourceModelSpec_RegistersDnsTarget` — cluster scope registers dns target with dnsfailure action and 3 flags
- `TestNewExpController_ReturnsClusterName` — controller name returns "cluster"
- `TestSelfExpModelCommandSpec_Metadata` — dns target Name/ShortDesc/LongDesc/Example completeness
- `TestClusterDnsFailureActionSpec_Metadata` — action Name/Aliases/ShortDesc/LongDesc, confirms LongDesc mentions `--dns-deployment`

**Boundary values (3):**
- `TestReadOriginalReplicas_BoundaryValues` — int32 max/min, overflow wrap-around, float, whitespace, hex (6 cases)
- `TestResolveDnsFlags_WhitespaceOnlyDeployment` — whitespace-only `--dns-deployment` trimmed to empty
- `TestResolveDnsFlags_OnlyServiceSet` — service-only defaults for namespace/deployment

**Exec-level error paths (3):**
- `TestExec_Create_ServiceNotFound_ReturnsFailStatus`
- `TestExec_Destroy_ServiceNotFound_ReturnsFailStatus`
- `TestExec_Create_ExplicitDeploymentNotFound`

**Behavioral coverage (4):**
- `TestInjectClusterDnsFailure_AlreadyAtZeroReplicas` — inject→restore with 0-replica Deployment
- `TestFindDnsDeployment_SupersetLabelsMatch` — selector as subset of pod template labels (real CoreDNS scenario)
- `TestDnsIdentifierHelpers_EmptyExplicitDeployment` — zero value identifier format
- `TestFindDnsDeployment_EmptyMapSelector` — empty map vs nil selector distinction

</details>

<details>
<summary><strong>Supplemental Pod DNS Tests (16 functions)</strong></summary>

**Registration/wiring (2):**
- `TestNewSelfExpModelCommandSpec_IncludesDnsFailureAction`
- `TestPodDnsFailureActionSpec_Metadata` — Name/Aliases/ShortDesc/LongDesc/Flags, LongDesc mentions UnreachableDnsNameserver

**Edge cases (2):**
- `TestPodUsesNameserver_EdgeCases` — DNSNone+nil config, empty nameservers, empty string match, multi-nameserver, searches-only (5 cases)
- `TestBeginPodDnsFailure_WithSearchesAndOptions` — full DNSConfig serialization with Searches/Options

**Multi-pod dedup (1):**
- `TestExec_Create_MultiPodSameWorkload_DedupWorks` — two pods, same Deployment, workload modified once

**Owner resolution (1):**
- `TestResolveTopLevelWorkload_ReplicaSetWithNilOwner` — standalone RS correctly rejected

**All workload kinds (1):**
- `TestInjectWorkloadDnsFailure_AllKinds` — parameterized Deployment/DaemonSet/StatefulSet injection

**Context missing (2):**
- `TestExec_Create_MissingContainerMeta_ReturnsError`
- `TestExec_Destroy_MissingContainerMeta_ReturnsError`

**DaemonSet/StatefulSet round trips (2):**
- `TestExec_CreateThenDestroy_DaemonSetRoundTrip`
- `TestExec_CreateThenDestroy_StatefulSetRoundTrip`

**Other (5):**
- `TestSupportedWorkloadKinds_ExactSet` — exactly 3 kinds
- `TestEndPodDnsFailure_WrongActionMarker_IsNoOp`
- `TestConstants` — constant values verification

</details>

---

*Generated by [pr-test-gen](https://github.com/chaosblade-io/chaosblade-operator/pull/303) — automated PR test analysis*

**Review by xcaspar — APPROVED** [review]
